### PR TITLE
feat: add deep health check for scheduler

### PR DESCRIPTION
**Pull request by Chethan-rao**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This PR adds support for performing a deeper health check of the scheduler which includes:

Database connection check with read, write and delete queries (Added support for database transaction too)
Redis connection check with set, get and delete key (Set key with expiry of 30sec)

The deep health check API returns a 200 status code if all the components are in good health, but returns 500 if at least one component's health is down

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

1. `/health`

image

2. `/health/ready`

image

3. `'health/ready` with some db error

image

Cannot be tested on sandbox as the endpoint is not exposed

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by NishantJoshi00 — APPROVED** [review]

**Review by SanchithHegde — DISMISSED** [review]

File: crates/router/src/bin/scheduler.rs [hunk 1]
```diff
@@ -37,7 +40,7 @@ async fn main() -> CustomResult<(), ProcessTrackerError> {
     // channel for listening to redis disconnect events
     let (redis_shutdown_signal_tx, redis_shutdown_signal_rx) = oneshot::channel();
     let state = Box::pin(routes::AppState::new(
-        conf,
+        conf.clone(),
```
Comment (NishantJoshi00) [inline crates/router/src/bin/scheduler.rs:43]: why was this clone required?

File: crates/router/src/bin/scheduler.rs [hunk 2]
```diff
@@ -68,6 +71,16 @@ async fn main() -> CustomResult<(), ProcessTrackerError> {
...
 
+    #[allow(clippy::expect_used)]
+    let web_server = Box::pin(start_web_server(
+        state.clone(),
+        scheduler_flow_str.to_string(),
+    ))
+    .await
+    .expect("Failed to create the server");
+
+    tokio::spawn(web_server);
```
Comment (NishantJoshi00) [inline crates/router/src/bin/scheduler.rs:82]: Can we have something like
```suggestion
    tokio::spawn(move async {
    	web_server.await;
    	logger::error!("The health check probe stopped working!");
    });
```

**Review by NishantJoshi00 — CHANGES_REQUESTED** [review]

File: crates/router/src/bin/scheduler.rs (hunk 1 above)
Comment (Chethan-rao) [reply crates/router/src/bin/scheduler.rs:43]: not required. removing

File: crates/router/src/bin/scheduler.rs [hunk 3]
```diff
@@ -76,6 +96,92 @@ async fn main() -> CustomResult<(), ProcessTrackerError> {
     Ok(())
 }
 
+pub async fn start_web_server(
+    state: routes::AppState,
+    service: String,
+) -> errors::ApplicationResult<Server> {
+    let server = state.conf.server.clone();
```
Comment (dracarys18) [inline crates/router/src/bin/scheduler.rs:103]: We should create a seperate config for `scheduler` server.

File: crates/router/src/bin/scheduler.rs [hunk 4]
```diff
@@ -76,6 +96,92 @@ async fn main() -> CustomResult<(), ProcessTrackerError> {
...
+    let web_server = actix_web::HttpServer::new(move || {
+        actix_web::App::new().service(Health::server(state.clone(), service.clone()))
+    })
+    .bind((server.host.as_str(), server.port))?
```
Comment (dracarys18) [inline crates/router/src/bin/scheduler.rs:107]: Make the workers configurable. And keep the default to 1

```rust
workers(conf.scheduler_server.workers)
```

File: crates/router/src/bin/scheduler.rs [hunk 5]
```diff
@@ -76,6 +96,92 @@ async fn main() -> CustomResult<(), ProcessTrackerError> {
...
+
+    Ok(web_server)
+}
+
+pub struct Health;
+
+impl Health {
+    pub fn server(state: routes::AppState, service: String) -> Scope {
+        web::scope("health")
+            .app_data(web::Data::new(state))
```
Comment (dracarys18) [inline crates/router/src/bin/scheduler.rs:119]: Why do we need the whole app state of `router` here. We need a smaller appstate with only necessary stuff.

File: crates/router/src/bin/scheduler.rs [hunk 6]
```diff
@@ -76,6 +96,92 @@ async fn main() -> CustomResult<(), ProcessTrackerError> {
...
+}
+#[instrument(skip_all)]
+pub async fn deep_health_check(
+    state: web::Data<routes::AppState>,
+    service: web::Data<String>,
+) -> impl actix_web::Responder {
+    let report = deep_health_check_func(state, service).await;
+    match report {
+        Ok(response) => {
+            services::http_response_json(serde_json::to_string(&response).unwrap_or_default())
```
Comment (dracarys18) [inline crates/router/src/bin/scheduler.rs:139]: Remove `unwrap_or_default` and handle errors properly

Comment (dracarys18) [inline crates/router/src/bin/scheduler.rs:139]: Instead you can use `ApplicationResponse::Json` as well

**Review by dracarys18 — CHANGES_REQUESTED** [review]

File: crates/router/src/bin/scheduler.rs (hunk 5 above)
Comment (Chethan-rao) [reply crates/router/src/bin/scheduler.rs:119]: The `HealthCheckInterface` is implemented on AppState itself so cannot ignore it. Also it was present in scheduler before

**Review by dracarys18 — APPROVED** [review]
LGTM

File: config/config.example.toml [hunk 7]
```diff
@@ -298,6 +298,11 @@ lower_fetch_limit = 1800          # Lower limit for fetching entries from redis
 lock_key = "PRODUCER_LOCKING_KEY" # The following keys defines the producer lock that is created in redis with
 lock_ttl = 160                    # the ttl being the expiry (in seconds)
 
+[scheduler.server]
+port = 3000
+host = "127.0.0.1"
+workers = 1
```
Comment (SanchithHegde) [inline config/config.example.toml:304]: Would be preferable if you can add comments about the same here, and in the `deployment/scheduler/{consumer,producer}.toml` files.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

**Review by SanchithHegde — APPROVED** [review]

**Review by dracarys18 — APPROVED** [review]

**Review by NishantJoshi00 — APPROVED** [review]

**Pull request merged**